ocaml/parsing/builtin_attributes.ml

```diff
@@ -935,20 +953,49 @@ let parse_zero_alloc_attribute ~is_arity_allowed ~default_arity attr =
         match filter_arity payload with
         | None -> default_arity, payload
         | Some (user_arity, payload) ->
-          if is_arity_allowed then
+          if in_signature then
             user_arity, payload
           else
             (warn_payload loc txt
                "The \"arity\" field is only supported on \"zero_alloc\" in \
                 signatures";
              default_arity, payload)
       in
-      parse_zero_alloc_payload ~loc ~arity ~warn ~empty:(empty arity) payload
+      let parse p =
+        parse_zero_alloc_payload ~loc ~arity ~warn ~empty:(empty arity) p
+      in
+      match filter_assume_unless_opt payload with
+      | None -> parse payload
+      | Some rest ->
+        if in_signature then
+          (warn_payload loc txt
+             "The payload \"assume_unless_opt\" is not supported \
+              in signatures.";
+           (* Treat [@zero_alloc assume_unless_opt] as [@zero_alloc] in signatures. *)
+           parse rest)
+        else
+          let no_other_payload = List.compare_length_with rest 0 = 0 in
+          if is_zero_alloc_check_enabled ~opt:true && no_other_payload then
+            (if on_application then
+               (* Treat is if there is no attribute.
+                  Check is not allowed on applications. *)
+               Default_zero_alloc
+             else
+               (* Treat [@zero_alloc assume_unless_opt] as [@zero_alloc],
+                  forcing the function to be checked.
+                  Setting [opt = false] to satisfy [@zero_alloc]
+                  and not only [@zero_alloc opt] on the corresponding signatures. *)
+               empty arity)
+          else
+            (* Treat "assume_unless_opt" as "assume".
+               Reuse standard parsing for better error messages. *)
+            parse payload
```

Comment:
The logic here is subtle. I spent a while thinking this was wrong, because we reach this case if the opt check is enabled but `no_other_payload` is false, and the comment says we're going to treat `assume_unless_opt` as assume in that case even though the opt check is on. If I understand correctly, it's actually fine, because the table has no entries for `assume_unless_opt` that include other things, so actually the table lookup will fail and we'll get default, not assume.  

This seems a bit error prone - if we decide in the future to support `assume_unless_opt` in combination with some other modifier, we then have a subtle bug here.  I think it would be better to just make the behavior explicit by returning default and calling `warn` when `no_other_payload` is true.